### DISCUSSION TITLE
How can i authenticate from physical SAM and to acess the data ?

### DISCUSSION THREAD

**Discussion by AmoledK** [metrodroid/metrodroid#832] [General]
I have a Physical SAM having Key for the authentication of Mifare Desire EV1 Card. How can i do this in current project?

**Comment by micolous** [comment]
I have no idea where to start on this, because I'm not familiar with these protocols. In terms of the problem you're trying to solve, Metrodroid is at the *very end of the chain* (ie: the very *last* step).

**As for where to start,** you might find Proxmark3's communities (or one of its forks) helpful for the ground work needed to talk to the SAM.

----

**Once you're past that point,** Metrodroid has [an API for custom DESFire authenticators][0] (we need it [for Leap card][1]), but it doesn't support communicating directly (or indirectly) with arbitrary SAMs.

[0]: https://github.com/metrodroid/metrodroid/blob/ed89fed9952fe424a7b5c8389ba52964b3634a1f/src/commonMain/kotlin/au/id/micolous/metrodroid/card/desfire/DesfireUnlocker.kt
[1]: https://github.com/metrodroid/metrodroid/blob/ed89fed9952fe424a7b5c8389ba52964b3634a1f/src/commonMain/kotlin/au/id/micolous/metrodroid/transit/tfi_leap/LeapUnlocker.kt

I presume the form-factor here is an ISO/IEC 7816-3 contact smartcard interface; in which case it's not clear to me how you'd actually expose the SAM to an Android or iOS device, as most of them don't actually have a spare smartcard slot.  So you'd need to:

* use an Android device which has SAM slots available;

  This is expensive, and would be _extremely_ niche – you would likely be the only person who could _ever_ use it. I have a strong preference against this being the _only_ way to use a SAM.

* use the jvmcli version of Metrodroid to interface with the SAM over PC/SC;

  This would still be pretty niche, but at least you should be able to use ordinary smartcard readers for this.

* write a protocol and a web service of your own (similar to Leap's) which can do the challenge/response parts;

  There would be data protection issues to work out if it's something which gated any of Metrodroid's functionality.

  One of my goals for Metrodroid is that it should work *totally* offline - so I don't like gating any functionality behind web services, even if it is something _I_ was hosting.  These cards have sensitive information on them, and the easiest way to handle data protection issues is by not collecting any data at all. 😅 

Based on the limited information I have at the moment, it's quite likely you will be literally the only person who can _ever_ use this feature, and it will require code changes, so I will use a well-worn line in open source: _patches welcomed™!_

I'd _also_ ask you to find a reasonably-priced small-quantity supplier of similar DESFire SAMs/cards which we can use as testing samples for any patches, because we'd want to make sure this keeps working.

**Reply by AmoledK** [reply]
As of now i want to do testing in my current device only. I have a device which can support SAM reading.  
and the Mifare Tag is compaitable with this particular SAM.
